x/stake/handler.go

```diff
@@ -0,0 +1,318 @@
+		return ErrBadBondingDenom().Result()
+	}
+	if ctx.IsCheckTx() {
+		return sdk.Result{
+			GasUsed: GasDeclareCandidacy,
+		}
+	}
+
+	candidate := NewCandidate(msg.CandidateAddr, msg.PubKey, msg.Description)
+	k.setCandidate(ctx, candidate)
+
+	// move coins from the msg.Address account to a (self-bond) delegator account
+	// the candidate account and global shares are updated within here
+	return k.delegateWithCandidate(ctx, msg.CandidateAddr, msg.Bond, candidate).Result()
+}
+
+func (k Keeper) handleMsgEditCandidacy(ctx sdk.Context, msg MsgEditCandidacy) sdk.Result {
+
+	// candidate must already be registered
+	candidate, found := k.GetCandidate(ctx, msg.CandidateAddr)
+	if !found {
+		return ErrBadCandidateAddr().Result()
+	}
+	if ctx.IsCheckTx() {
+		return sdk.Result{
+			GasUsed: GasEditCandidacy,
+		}
+	}
+	if candidate.Status == Unbonded { //candidate has been withdrawn
+		return ErrBondNotNominated().Result()
+	}
+
+	// XXX move to types
+	//check and edit any of the editable terms
+	if msg.Description.Moniker != "" {
+		candidate.Description.Moniker = msg.Description.Moniker
+	}
+	if msg.Description.Identity != "" {
+		candidate.Description.Identity = msg.Description.Identity
+	}
+	if msg.Description.Website != "" {
+		candidate.Description.Website = msg.Description.Website
+	}
+	if msg.Description.Details != "" {
+		candidate.Description.Details = msg.Description.Details
+	}
+
+	k.setCandidate(ctx, candidate)
+	return sdk.Result{}
+}
+
+func (k Keeper) handleMsgDelegate(ctx sdk.Context, msg MsgDelegate) sdk.Result {
+
+	candidate, found := k.GetCandidate(ctx, msg.CandidateAddr)
+	if !found {
+		return ErrBadCandidateAddr().Result()
+	}
+	if msg.Bond.Denom != k.GetParams(ctx).BondDenom {
+		return ErrBadBondingDenom().Result()
+	}
+	if ctx.IsCheckTx() {
+		return sdk.Result{
+			GasUsed: GasDelegate,
+		}
+	}
+	return k.delegateWithCandidate(ctx, msg.DelegatorAddr, msg.Bond, candidate).Result()
+}
+
+func (k Keeper) delegateWithCandidate(ctx sdk.Context, delegatorAddr sdk.Address,
+	bondAmt sdk.Coin, candidate Candidate) sdk.Error {
+
+	if candidate.Status == Revoked { //candidate has been withdrawn
+		return ErrBondNotNominated()
+	}
+
+	// Get or create the delegator bond
+	existingBond, found := k.getDelegatorBond(ctx, delegatorAddr, candidate.Address)
+	if !found {
+		existingBond = DelegatorBond{
+			DelegatorAddr: delegatorAddr,
+			CandidateAddr: candidate.Address,
+			Shares:        sdk.ZeroRat,
+		}
+	}
+
+	// Account new shares, save
+	err := k.BondCoins(ctx, existingBond, candidate, bondAmt)
+	if err != nil {
+		return err
+	}
+	k.setDelegatorBond(ctx, existingBond)
+	k.setCandidate(ctx, candidate)
+	return nil
+}
+
+// Perform all the actions required to bond tokens to a delegator bond from their account
+func (k Keeper) BondCoins(ctx sdk.Context, bond DelegatorBond, candidate Candidate, amount sdk.Coin) sdk.Error {
+
+	_, err := k.coinKeeper.SubtractCoins(ctx, bond.DelegatorAddr, sdk.Coins{amount})
```

Comment:
Spec says coins should be transferred to pool account - is that account not necessary to track explicitly?

Reply:
Discussed; spec should change @milosevic - no longer a need for a pool account.